core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java

```diff
@@ -45,6 +55,14 @@
     private final PlannerContext plannerContext;
     private final StatementAnalyzerFactory statementAnalyzerFactory;
 
+    private final NonEvictableCache<CacheIdentityKey, ExpressionAnalyzer> analyzerCache = buildNonEvictableCache(
```

Comment:
`ExpressionAnalyzer` is a ton of mutable state, it's totally insuitable for caching & sharing.

Reply:
That was proposed by @sopel39 . My previous version was actually caching only `Map<RefNode<Expression>, Type>`

Reply:
`Type` is immutable, so it was definitely a better idea 

(not judging whether this particular cache in this place is required)

Reply:
@sopel39 wdyt?

Reply:
> ExpressionAnalyzer is a ton of mutable state, it's totally insuitable for caching & sharing.

I don't think it's a problem since it only adds new information to analysis. It already caches via `io.trino.sql.analyzer.ExpressionAnalyzer.Visitor#process`. Since we only care about types, we could maybe prime new `ExpressionAnalyzer` instance with `Map<NodeRef<Expression>, Type>`, but I'm not sure it's worth extra complexity.

> That was proposed by @sopel39 . My previous version was actually caching only Map<RefNode<Expression>, Type>

IIRC that map was not passed to `ExpressionAnalyzer` hence it wouldn't work with sub-expressions

Reply:
Anyway, for me planning is a process so keeping `ExpressionAnalyzer` is not really an issue. Actually, it's probably desirable since I want to cache `ExpressionInterpreter` too and it needs to be able to fetch type information for new expressions.

Reply:
I strongly opt for caching `Map<RefNode, Type>`, as that's what you actually need here.

The ExpressionAnalyzer was not designed to be created in multiple instances throughout the Planner. It is intended for the Analysis phase, where it provides different kind of information to the Planner via the Analysis object. If you need to analyze expressions later for some reason, there's the `analyzeExpressions` method, and `createConstantAnalyzer` method, depending on your use case. Those methods give you access to the "analyzing" capability of the `ExpressionAnalyzer` while they hide the complexity which is not relevant at that point.

Reply:
> The ExpressionAnalyzer was not designed to be created in multiple instances throughout the Planner.

This is exactly what is happening right now. Every `io.trino.sql.planner.TypeAnalyzer#getTypes` call creates a new `ExpressionAnalyzer` instance (via `analyzeExpressions`) and does a full analysis. Overall it's expensive if done repeatedly.
We want to reduce that cost by keeping `ExpressionAnalyzer` instance during planning, which should be fine since planning is a process rather than moving from one immutable state to the other.

> I strongly opt for caching Map<RefNode, Type>, as that's what you actually need here.

I would like to keep `ExpressionInterpreter` too. For that I need a utility to return `type` for a (sub)expression rather than analyzing entire expression to get full `Map<NodeRef<Expression>, Type>` map. In that context keeping `ExpressionAnalyzer` is more natural.